### PR TITLE
AUT-115 Borrower endpoints

### PR DESCRIPTION
**Pull request by azayatsfv**
added borrowers endpoints:

GET get_borrower
GET get_customer_loan_list
GET get_customers
POST add_borrower_pair
POST add_consent
POST delete_borrower_pair
POST set_customer
POST set_person_data

### CODE REVIEW

File: APIs/incomes/requests/set_incomes.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def build_payload(self) -> Dict[str, List[Dict[Any, Union[List[Dict[str, Any]],
     obj = SetIncomesRequest(loan_number_id=10000001, payload_dict=None, session_id=123456, nonce=123245687, pretty_print=False, **kwargs)
     print(f"\nPAYLOAD: {pprint.pformat(obj.payload)}")
 
-    print("\nTesting SetIncomesRequest - payload_dict")
+    print("\nTesting SetCustomerRequest - payload_dict")
```
Comment (darkhipenko) [inline APIs/incomes/requests/set_incomes.py:88]: SetIncomesRequest was replaced, need change back

File: APIs/persons/borrowers/client.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+class BorrowersClient(BaseClient):
+    CONTENT_TYPE = "Content-Type"
+    APPLICATION_JSON = "application/json"
+    json_headers = {CONTENT_TYPE: APPLICATION_JSON
+    }
+
+    def add_borrower_pair(self, loan_number_id, session_id=None, nonce=None, pretty_print=False):
+        request_model = LoanNumberIdRequestModel(loan_number_id = loan_number_id,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print)
+        return self.get(resource_endpoint = ApiEndpoints.ADD_BORROWER_PAIR, response_model = AddBorrowerPairResponse,
```
Comment (darkhipenko) [inline APIs/persons/borrowers/client.py:37]: POST instead of GET

File: APIs/persons/borrowers/client.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        request_model = DeleteBorrowerPairRequest(loan_number_id = loan_number_id,
+            primary_borrower_id = primary_borrower_id, secondary_borrower_id = secondary_borrower_id,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print)
+        return self.post(resource_endpoint = ApiEndpoints.DELETE_BORROWER_PAIR, response_model = CommonResponse,
+            params = request_model.as_params_dict, data = request_model.payload, headers = self.json_headers)
+
+    def get_customer_loan_list(self, loan_number_id, session_id=None, nonce=None, pretty_print=False):
+        request_model = LoanNumberIdRequestModel(loan_number_id = loan_number_id,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print)
+        return self.post(resource_endpoint = ApiEndpoints.GET_CUSTOMER_LOAN_LIST,
```
Comment (darkhipenko) [inline APIs/persons/borrowers/client.py:60]: GET instead of POST

File: APIs/persons/borrowers/client.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    def get_borrower(self, loan_number_id, session_id=None, nonce=None, pretty_print=False):
+        request_model = LoanNumberIdRequestModel(loan_number_id = loan_number_id,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print)
+        return self.post(resource_endpoint = ApiEndpoints.GET_BORROWER, response_model = GetBorrowerResponse,
+            params = request_model.as_params_dict, data = request_model.payload)
+
+    def get_customers(self, loan_number_id, session_id=None, nonce=None, pretty_print=False):
+        request_model = LoanNumberIdRequestModel(loan_number_id = loan_number_id,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print)
+        return self.post(resource_endpoint = ApiEndpoints.GET_CUSTOMERS, response_model = GetCustomersResponse,
```
Comment (darkhipenko) [inline APIs/persons/borrowers/client.py:73]: GET instead of POST

File: APIs/persons/borrowers/client.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            params = request_model.as_params_dict, data = request_model.payload)
+
+    def set_customer(self, loan_number_id, payload_dict=None, session_id=None, nonce=None, pretty_print=False,
+            **kwargs):
+        request_model = SetCustomerRequest(loan_number_id = loan_number_id, payload_dict = payload_dict,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print,
+            **kwargs)
+        return self.post(resource_endpoint = ApiEndpoints.SET_CUSTOMER, response_model = CommonResponse,
```
Comment (darkhipenko) [inline APIs/persons/borrowers/client.py:81]: headers=json

File: APIs/persons/borrowers/client.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            params = request_model.as_params_dict, data = request_model.payload)
+
+    def set_person_data(self, loan_number_id, payload_dict=None, session_id=None, nonce=None, pretty_print=False,
+            **kwargs):
+        request_model = SetPersonDataRequest(loan_number_id = loan_number_id, payload_dict = payload_dict,
+            session_id = self._get_session_id(session_id), nonce = self._get_nonce(nonce), pretty_print = pretty_print,
+            **kwargs)
+        return self.post(resource_endpoint = ApiEndpoints.SET_PERSON_DATA, response_model = CommonResponse,
+            params = request_model.as_params_dict, data = request_model.payload)
```
Comment (darkhipenko) [inline APIs/persons/borrowers/client.py:90]: headers=json

File: APIs/persons/borrowers/models/borrower_pair.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+@dataclass
+class BorrowerPairsInfoKeys:
+    PRIMARY_BORROWER_ID: str = "PrimaryBorrowerId"
+    SECONDARY_BORROWER_ID: str = "SecondaryBorrowerId"
+
+class BorrowerPair(BaseResponse):
+    _ADD_KEYS = [field.default for field in fields(BorrowerPairsInfoKeys)]
+
+class BorrowerPairs(BaseListResponse):
+    _SUB_MODEL = BorrowerPair
```
Comment (darkhipenko) [inline APIs/persons/borrowers/models/borrower_pair.py:13]: BorrowerPairsInfoKeys will be used in response, you just use 2 attributes there and remove this code

File: APIs/persons/borrowers/requests/add_borrower_pair.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+class AddBorrowerPairRequest(SimpleRequestModel):
+    def __init__(self, loan_number_id, session_id, nonce, pretty_print):
+        self.loan_number_id = loan_number_id
+        super().__init__(session_id=session_id, nonce=nonce, pretty_print=pretty_print)
+
+    def to_params(self):
+        args = super().to_params()
+        args[AddBorrowerPairRequestParams.LOAN_NUMBER_ID] = self.loan_number_id
+        return args
```
Comment (darkhipenko) [inline APIs/persons/borrowers/requests/add_borrower_pair.py:18]: Can this file be replaced just by using class LoanNumberIdRequestModel in the request?

File: APIs/persons/borrowers/requests/get_borrower.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+class GetBorrowerRequest(SimpleRequestModel):
+    def __init__(self, loan_number_id, session_id, nonce, pretty_print):
+        self.loan_number_id = loan_number_id
+        super().__init__(session_id=session_id, nonce=nonce, pretty_print=pretty_print)
+
+    def to_params(self):
+        args = super().to_params()
+        args[GetBorrowerRequestParams.LOAN_NUMBER_ID] = self.loan_number_id
+        return args
```
Comment (darkhipenko) [inline APIs/persons/borrowers/requests/get_borrower.py:18]: Can this file be replaced just by using class LoanNumberIdRequestModel in the request?

File: APIs/persons/borrowers/requests/get_customer_loan_list.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+class GetCustomerLoanListRequest(SimpleRequestModel):
+    def __init__(self, loan_number_id, session_id, nonce, pretty_print):
+        self.loan_number_id = loan_number_id
+        super().__init__(session_id=session_id, nonce=nonce, pretty_print=pretty_print)
+
+    def to_params(self):
+        args = super().to_params()
+        args[GetCustomerLoanListRequestParams.LOAN_NUMBER_ID] = self.loan_number_id
+        return args
```
Comment (darkhipenko) [inline APIs/persons/borrowers/requests/get_customer_loan_list.py:18]: Can this file be replaced just by using class LoanNumberIdRequestModel in the request?

File: APIs/persons/borrowers/requests/get_customers.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+class GetCustomersRequest(SimpleRequestModel):
+    def __init__(self, loan_number_id, session_id, nonce, pretty_print):
+        self.loan_number_id = loan_number_id
+        super().__init__(session_id=session_id, nonce=nonce, pretty_print=pretty_print)
+
+    def to_params(self):
+        args = super().to_params()
+        args[GetCustomersRequestParams.LOAN_NUMBER_ID] = self.loan_number_id
+        return args
```
Comment (darkhipenko) [inline APIs/persons/borrowers/requests/get_customers.py:18]: Can this file be replaced just by using class LoanNumberIdRequestModel in the request?

File: tests/test_person_borrower.py [hunk 12]
```diff
@@ -1,53 +1,41 @@
...
-    LoanListRowValuesList, LoanListRowColValueEntryKeys, LoanListRowCol, LoanListRowColList, LoanListRowsKeys,
-    LoanListHeaderColumnKeys, CustomerLoanListKeys, CustomerLoanList
-)
-from APIs.person.borrower.responses.get_borrower import GetBorrower
-from APIs.person.borrower.responses.get_customers import GetCustomers
-from APIs.person.borrower.responses.get_customer_loan_list import GetCustomerLoanList
-from APIs.person.borrower.responses.set_customer import SetCustomer
-from APIs.person.borrower.models.customers import Customer, CustomerKeys, CustomerList, CustomerListKeys
+from APIs.persons.borrowers.models.borrower import Borrower, BorrowersInfoKeys, Borrowers, BorrowersKeys
+from APIs.persons.borrowers.models.loan_list import (CustomerLoanListsInfoKeys,CustomerLoanListsKeys, CustomerLoanList)
```
Comment (darkhipenko) [inline tests/test_person_borrower.py:5]: parenthesis are not needed

File: tests/test_person_borrower.py [hunk 13]
```diff
@@ -117,71 +93,19 @@ def test_SetCustomer_response(self):
...
-            self._validate_response(model=sub_model, model_data=data[index])
-
-    def test_customer_loan_list_model(self):
-        header_data = [build_loan_list_entry() for _ in range(NUMBER_OF_HEADERS)]
-        row_data = [{LoanListRowColValueEntryKeys.COL:
-                    [build_loan_list_row_entry() for _ in range(NUMBER_OF_HEADERS)]} for _ in range(NUMBER_OF_ROWS)]
-
-        data = {LoanListHeaderColumnKeys.COLS: header_data,
-                LoanListRowsKeys.ROWS: row_data}
+        data = build_customer_loan_list()
```
Comment (darkhipenko) [inline tests/test_person_borrower.py:96]: tests on lines 120-171 were deleted

File: tests/test_person_borrower.py [hunk 14]
```diff
@@ -117,71 +93,19 @@ def test_SetCustomer_response(self):
...
-        response = GetCustomerLoanList(**args)
-        self._validate_response(model=response, model_data=args)
+        key = CustomerLoanListsKeys.CUSTOMER_LOANS
+        data = response_args.copy()
+        data[key] = [build_customer_loan_list() for _ in range(NUMBER_OF_BORROWERS)]
+        response = GetCustomerLoanListResponse(**data)
+
+        self._verify(f"{response.model_name}: Verify response has '{key}' attribute.", actual = hasattr(response, key),
+            expected = True)
+        self._validate_response(model = response, model_data = data)
```
Comment (darkhipenko) [inline tests/test_person_borrower.py:108]: Not sure how this test works

File: tests/test_person_borrower.py (hunk 13 above)
Comment (azayatsfv) [reply tests/test_person_borrower.py:96]: deleted, because response structure was changed

**Review by darkhipenko — APPROVED** [review]
Approved

**Pull request merged**